### PR TITLE
TF-2102 [Part 1] Implement restore deleted message API

### PR DESCRIPTION
**Pull request by hieutbui**
## Issue
---
#2102 

### DOD
---
- [x] Restore deleted message success
 
## Resolved
---

https://github.com/linagora/tmail-flutter/assets/80142234/c2a91ec9-e3dd-40c9-9106-dcaa91bad1c5

### CODE REVIEW

**Comment by dab246** [comment]
Screenshot 2023-12-20 at 18 53 35

Why has the recovery been successful but there are still these mailboxes in the Trash?

**Review by hoangdat — APPROVED** [review]

File: lib/features/email/data/network/email_api.dart [hunk 1]
```diff
@@ -687,4 +694,52 @@ class EmailAPI with HandleSetErrorMixin {
...
+          getEmailRecoveryActionInvocation.methodCallId,
+          GetEmailRecoveryActionResponse.deserialize
+        );
+
+        if (resultList?.list.isNotEmpty == true) {
+          return resultList!.list.firstWhere((element) => element.id == emailRecoveryActionId);
+        } else {
+          throw NotFoundEmailRecoveryActionException();
+        }
+      });
```
Comment (dab246) [inline lib/features/email/data/network/email_api.dart:743]: We should unify the way we write code in handling api from jmap. Please make like the following image

Screenshot 2024-01-02 at 10 44 22

File: lib/features/email/domain/state/get_restored_deleted_message_state.dart [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+import 'package:email_recovery/email_recovery/email_recovery_action.dart';
+
+class GetRestoredDeletedMessageLoading extends UIState {}
```
Comment (dab246) [inline lib/features/email/domain/state/get_restored_deleted_message_state.dart:5]: Should be extended from LoadingState

File: lib/features/email/domain/state/restore_deleted_message_state.dart [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+import 'package:email_recovery/email_recovery/email_recovery_action.dart';
+
+class RestoreDeletedMessageLoading extends UIState {}
```
Comment (dab246) [inline lib/features/email/domain/state/restore_deleted_message_state.dart:5]: idem

File: lib/features/email/domain/usecases/get_restored_deleted_message_interactor.dart [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  GetRestoredDeletedMessageInterator(this._emailRepository);
+
+  Stream<Either<Failure, Success>> execute(
+    EmailRecoveryActionId emailRecoveryActionId,
+    StreamController<Either<Failure, Success>> onProgressController
+  ) async* {
+    try {
+      yield Right<Failure, Success>(GetRestoredDeletedMessageLoading());
+      onProgressController.add(Right(GetRestoredDeletedMessageLoading()));
```
Comment (dab246) [inline lib/features/email/domain/usecases/get_restored_deleted_message_interactor.dart:21]: We should handle it in the controller. It is very difficult to write unit tests later

File: lib/features/email/domain/usecases/restore_deleted_message_interactor.dart.dart [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  Stream<Either<Failure, Success>> execute(
+    RestoredDeletedMessageRequest newRecoveryRequest,
+    StreamController<Either<Failure, Success>> onProgressController
+  ) async* {
+    try {
+      yield Right<Failure, Success>(RestoreDeletedMessageLoading());
+      onProgressController.add(Right(RestoreDeletedMessageLoading()));
+      final emailRecovery = await _emailRepository.restoreDeletedMessage(newRecoveryRequest);
+      yield Right<Failure, Success>(RestoreDeletedMessageSuccess(emailRecovery));
+      onProgressController.add(Right(RestoreDeletedMessageSuccess(emailRecovery)));
```
Comment (dab246) [inline lib/features/email/domain/usecases/restore_deleted_message_interactor.dart.dart:24]: idem

**Comment by dab246** [comment]
@hieutbui Please rebase it

**Review by dab246 — APPROVED** [review]

**Pull request merged**